components/canvas/webgl_thread.rs

```diff
@@ -0,0 +1,1191 @@
+            ctx.gl().wait_sync(gl_sync, 0, gl::TIMEOUT_IGNORED);
+            // Release the GLSync object.
+            ctx.gl().delete_sync(gl_sync);
+        }
+    }
+
+    /// Creates a new WebGLContext
+    fn create_webgl_context(&mut self,
+                            size: Size2D<i32>,
+                            attributes: GLContextAttributes)
+                            -> Result<(WebGLContextId, GLLimits, WebGLContextShareMode), String> {
+        // First try to create a shared context for the best performance.
+        // Fallback to readback mode if the shared context creation fails.
+        let result = self.gl_factory.new_shared_context(size, attributes)
+                                    .map(|r| (r, WebGLContextShareMode::SharedTexture))
+                                    .or_else(|_| {
+                                        let ctx = self.gl_factory.new_context(size, attributes);
+                                        ctx.map(|r| (r, WebGLContextShareMode::Readback))
+                                    });
+
+        // Creating a new GLContext may make the current bound context_id dirty.
+        // Clear it to ensure that  make_current() is called in subsequent commands.
+        self.current_bound_webgl_context_id = None;
+
+        match result {
+            Ok((ctx, share_mode)) => {
+                let id = WebGLContextId(self.next_webgl_id);
+                let (size, texture_id, limits) = ctx.get_info();
+                self.next_webgl_id += 1;
+                self.contexts.insert(id, ctx);
+                self.cached_context_info.insert(id, WebGLContextInfo {
+                    texture_id,
+                    size,
+                    alpha: attributes.alpha,
+                    image_key: None,
+                    share_mode,
+                    gl_sync: None,
+                    old_image_key: None,
+                    very_old_image_key: None,
+                });
+
+                self.observer.on_context_create(id, texture_id, size);
+
+                Ok((id, limits, share_mode))
+            },
+            Err(msg) => {
+                Err(msg.to_owned())
+            }
+        }
+    }
+
+    /// Resizes a WebGLContext
+    fn resize_webgl_context(&mut self,
+                            context_id: WebGLContextId,
+                            size: Size2D<i32>,
+                            sender: WebGLSender<Result<(), String>>) {
+        let ctx = self.contexts.get_mut(&context_id).unwrap();
+        if Some(context_id) != self.current_bound_webgl_context_id {
+            ctx.make_current();
+            self.current_bound_webgl_context_id = Some(context_id);
+        }
+        match ctx.resize(size) {
+            Ok(_) => {
+                let (real_size, texture_id, _) = ctx.get_info();
+                self.observer.on_context_resize(context_id, texture_id, real_size);
+
+                let info = self.cached_context_info.get_mut(&context_id).unwrap();
```

Comment:
nit: Can use `self.cached_context_info[&context_id]`, here and in a bunch of other places, to get nicer error messages.